Microsoft.Toolkit.Uwp.UI.Controls/TextBoxMask/TextBoxMask.cs

```diff
@@ -0,0 +1,303 @@
+                        deleteBackspaceIndex++;
+                    }
+                }
+            }
+
+            // case adding data at the end of the textbox
+            if (oldSelectionStart >= oldText.Length && !isDeleteOrBackspace)
+            {
+                textbox.Text = oldText;
+                if (oldText.Length >= 0)
+                {
+                    textbox.SelectionStart = oldText.Length;
+                }
+
+                return;
+            }
+
+            var selectedChar = textbox.SelectionStart > 0 ?
+                textbox.Text[textbox.SelectionStart - 1] :
+                placeHolder;
+
+            var textArray = oldText.ToCharArray();
+
+            // detect if backspace or delete is triggered to handle the right removed character
+            var newSelectionIndex = oldSelectionStart - deleteBackspaceIndex;
+
+            // for handling single key click add +1 to match length for selection =1
+            var singleOrMultiSelectionIndex = oldSelectionLength == 0 ? oldSelectionLength + 1 : oldSelectionLength;
+            for (int i = newSelectionIndex; i < (oldSelectionStart - deleteBackspaceIndex + singleOrMultiSelectionIndex); i++)
+            {
+                var maskChar = mask[i];
+
+                // If dynamic character a,9,* or custom
+                if (representationDictionary.ContainsKey(maskChar))
+                {
+                    if (isDeleteOrBackspace)
+                    {
+                        textArray[i] = placeHolder;
+                        continue;
+                    }
+
+                    var pattern = representationDictionary[maskChar];
+                    if (Regex.IsMatch(selectedChar.ToString(), pattern))
+                    {
+                        textArray[i] = selectedChar;
+
+                        // updating text box new index
+                        newSelectionIndex++;
+                    }
+
+                    // character doesn't match the pattern get the old character
+                    else
+                    {
+                        // if single press don't change
+                        if (oldSelectionLength == 0)
+                        {
+                            textArray[i] = oldText[i];
+                        }
+
+                        // if change in selection reset to default place holder instead of keeping the old valid to be clear for the user
+                        else
+                        {
+                            textArray[i] = placeHolder;
+                        }
+                    }
+                }
+
+                // if fixed character
+                else
+                {
+                    textArray[i] = oldText[i];
+
+                    // updating text box new index
+                    newSelectionIndex++;
+                }
+            }
+
+            textbox.Text = new string(textArray);
+            textbox.SetValue(OldTextProperty, textbox.Text);
```

Comment:
We need on TextChanging event 2 values the last old text and the current one which is read from textbox.text and that's the only way I can get the Old right text